packages/react-native-executorch/common/rnexecutorch/models/speech_to_text/asr/ASR.cpp

```diff
@@ -0,0 +1,306 @@
+        generatedTokens[i] >= this->timestampBeginToken) {
+      int32_t start = prevTimestamp;
+      int32_t end = generatedTokens[i - 1];
+      auto words = this->estimateWordLevelTimestampsLinear(tokens, start, end);
+      if (words.size()) {
+        segments.push_back({words, 0.0});
+      }
+      tokens.clear();
+      prevTimestamp = generatedTokens[i];
+    }
+  }
+
+  int32_t start = prevTimestamp;
+  int32_t end = generatedTokens[generatedTokens.size() - 2];
+  auto words = this->estimateWordLevelTimestampsLinear(tokens, start, end);
+
+  if (words.size()) {
+    segments.emplace_back(Segment{words, 0.0});
+  }
+
+  float scalingFactor =
+      static_cast<float>(waveform.size()) / ASR::samplingRate /
+      ((end - this->timestampBeginToken) * ASR::timePrecision);
+  if (scalingFactor < 1.0f) {
+    for (auto &seg : segments) {
+      for (auto &w : seg.words) {
+        w.start *= scalingFactor;
+        w.end *= scalingFactor;
+      }
+    }
+  }
+
+  return segments;
+}
+
+std::vector<Word>
+ASR::estimateWordLevelTimestampsLinear(std::span<const int32_t> tokens,
+                                       int32_t start, int32_t end) const {
+  const std::vector<int32_t> tokensVec(tokens.begin(), tokens.end());
+  std::string segmentText = this->tokenizer.decode(tokensVec, true);
+  std::istringstream iss(segmentText);
+  std::vector<std::string> wordsStr;
+  std::string word;
+  while (iss >> word) {
+    wordsStr.push_back(" " + word);
+  }
+
+  int32_t numChars = 0;
+  for (auto &w : wordsStr)
+    numChars += static_cast<int32_t>(w.size());
+
+  float duration = (end - start) * ASR::timePrecision;
+  float timePerChar = duration / std::max(1, numChars);
+  float startOffset = (start - timestampBeginToken) * ASR::timePrecision;
+
+  std::vector<Word> wordObjs;
+  int prevCharCount = 0;
+  for (auto &w : wordsStr) {
+    float wStart = startOffset + prevCharCount * timePerChar;
+    float wEnd = wStart + timePerChar * w.size();
+    wordObjs.push_back({w, wStart, wEnd});
+    prevCharCount += w.size();
+  }
+
+  return wordObjs;
+}
+
+std::vector<Segment> ASR::transcribe(std::span<const float> waveform,
+                                     const DecodingOptions &options) const {
+  int32_t seek = 0;
+  std::vector<Segment> results;
+
+  while (seek * ASR::samplingRate < waveform.size()) {
+    int32_t start = seek * ASR::samplingRate;
+    int32_t end = std::min<int32_t>((seek + ASR::chunkSize) * ASR::samplingRate,
+                                    waveform.size());
+    std::span<const float> chunk = waveform.subspan(start, end - start);
+
+    if (chunk.size() < ASR::minChunkSamples) {
+      break;
+    }
+
+    auto segments = this->generateWithFallback(chunk, options);
+
+    for (auto &seg : segments) {
+      for (auto &w : seg.words) {
+        w.start += seek;
+        w.end += seek;
+      }
+    }
+
+    results.insert(results.end(), segments.begin(), segments.end());
```

Comment:
is `Segment` is slower to copy than move, consider inserting with moving inserter since you don't use `segments` anymore (but first move line `seek = static_cast<int>(segments.back().words.back().end);` before inserting)